### ISSUE TITLE
Asp.Net Core does not properly handle the error with polymorphic deserialization from STJ when type discriminator is not specified

### ISSUE THREAD

**Issue by ilya-scale** [dotnet/aspnetcore#54037]
### Description

When type discriminator is not specified in the request, the error is not handled and 500 is returned instead of 400.

Example request 1:
```json
{
	"type": "unknown",
	"someOtherProperty": "value"
}
```

Example request 2:
```json
{
	"someOtherProperty": "value"
}
```

Controller action:

```csharp
[HttpPost("test")]
public void Test(Base request)
{
        
}
```

Models:

```csharp
[JsonPolymorphic(TypeDiscriminatorPropertyName = "type")]
[JsonDerivedType(typeof(Derived), "derived")]
public abstract record Base
{
}

public record Derived : Base
{
}
```

### Reproduction Steps

1. Create a controller with the action and models from Description
2. Do a call with the example request 1, response is 400

```json
{
	"type": "https://tools.ietf.org/html/rfc9110#section-15.5.1",
	"title": "One or more validation errors occurred.",
	"status": 400,
	"errors": {
		"$": [
			"Read unrecognized type discriminator id 'unknown'. Path: $ | LineNumber: 2 | BytePositionInLine: 21."
		],
		"request": [
			"The request field is required."
		]
	},
	"traceId": "00-a4e0c60070a4c53222d0622acad8d198-e18068d6fe5d5773-00"
}
```

3. Do a call with the example request 2, response is 500
```
System.NotSupportedException: Deserialization of types without a parameterless constructor, a singular parameterized constructor, or a parameterized constructor annotated with 'JsonConstructorAttribute' is not supported.
```

### Expected behavior

Ideally I expect the 400 error that will say clearly that the "type" TypeDiscriminator has to be provided (since the base class is abstract it is clear it cannot be constructed).

At the very least some form of 400 with json that tells that json cannot be deserialized instead of 500 (there was nothing "wrong" with the server, but client provided a bad request which is clearly a 400).

### Actual behavior

A 500 is returned and the exception is not handled by asp.net core (unlike example 1 request)

### Known Workarounds

One can probably make class not abstract and do a validation oneself that the "base" class is not supported.

### Configuration

.Net 8
MacOS ARM

### Other information

It is somewhat related to the case when type discriminator is not the first field. If it is not the first - it will be the same error I think, which should also be handled ideally, but this can be fixed in .Net 9 hopefully.

**Comment by ilya-scale** [comment]
Yes, deserialization is of course not the same as validation, but deserialization is a part of handling the http input, and since that input is wrong (does not conform to the expected model) so that we cannot deserialize, I expect to have a 400 the same way as if the whole json is not properly formatted, etc. Failing with 500 seems a bit unnecessary since we know for fact this was a bad request: json that cannot be deserialized.

**Comment by zachsaw-lmg** [comment]
Just ran into this problem - been over a year and it's still untriaged :(

**Comment by halter73** [comment]
@eiriktsarpalis I think I agree with @zachsaw-lmg that this really ought to be a `JsonException` rather than a `NotSupportedException`. In any case where we throw a `NotSupportedException`, it should be easy to avoid either through prior knowledge about how the calling code is deployed/used or through appropriate runtime checks. I don't think either is the case here.

I think making this a `JsonException` would be more appropriate when bad data rather than bad configuration can at least possibly be the cause of the error, and that by itself would fix this error.

**Comment by eiriktsarpalis** [comment]
> I think making this a JsonException would be more appropriate when bad data rather than bad configuration

This is actually a case of bad configuration. You get the same error message if you run this code:

```C#
JsonSerializer.Deserialize<MyPoco>("{}"); // Cannot deserialize to abstract types

abstract class MyPoco;
```

The reason why it only happens with a specific payload is because STJ performs delayed validation of the deserialization constructor, which in this case triggers when a given type discriminator (or lack thereof) results in the abstract type being used.

I'm not a fan of delayed validation personally, and would much rather have all validations be performed eagerly for early feedback. The problem is that changing this today would result in breaking changes in cases where users serialize abstract types without care for deserializing them.

In other words, raising a 500 seems appropriate in this case.

**Comment by ilya-scale** [comment]
I do believe regardless of the internal workings of the STJ, throwing 500 is not correct as the only thing client did is "forgot" to send a field. Ideally the server should tell the client why this request will not be accepted rather than just die without any explanation (we e.g. hide the exception for security reasons since it is quite standard). 

So if the client does not understand what is going on, returning 500 would just create a lot of friction, rather than a correct 400 with an explanation that this field is required. 

I do not see how we can configure anything differently on backend either, since the configuration is all correct, we just cannot deserialize if the type is not provided, which is a simple request error.

**Comment by eiriktsarpalis** [comment]
> I do believe regardless of the internal workings of the STJ, throwing 500 is not correct as the only thing client did is "forgot" to send a field.

The type discriminator is not a required field though. Not specifying one makes the serializer default to the base type.

**Comment by ilya-scale** [comment]
It does not quite work if the base type is abstract, so if you suggest that the current solution is acceptable, that would mean that for every hierarchy that relies on abstract class on top, one has to change that abstract class to a concrete one and fail the validation always, by e.g. doing it via `IValidateableObject`.

That seems like a lot of unnecessary work.

**Comment by eiriktsarpalis** [comment]
> if you suggest that the current solution is acceptable

For better or for worse, this is the behavior that got shipped at the time. Type discriminators are optional because we had use cases where they needed to be optional. We can have a conversation about introducing a new flag that promotes them to a required field (and emit a relevant `JsonException` on deserialization) but this would have to be opt in to prevent breaking changes.

**Comment by ilya-scale** [comment]
For me an opt in flag would work fine.

Although perhaps changing the type of exception thrown and converting it to 400 would also work? That would not make the discriminator required by default, but only would return another response if it is not provided (and the base class cannot be constructed). 

Or would that be too much of a breaking change as well?

**Comment by halter73** [comment]
> For better or for worse, this is the behavior that got shipped at the time.

I'm really surprised by this pushback. Every bug fix is technically a breaking change. While I understand that System.Text.Json has a lot of users and making bar for what qualifies as a breaking change pretty low, I think changing this from a NotSupportedException to a JsonException falls well below the bar. We could make a breaking change announcement if we are really worried about it, but I'd be willing to bet that zero people are will actually be broken by changing this NotSupportedException to a JsonException.

@eiriktsarpalis Can you provide an example of the code you're worried about breaking if we make this change? I'm not saying it's impossible to come up with, but could you point to any open source code or internal code that would actually be broken?

Practically speaking, I think changing a NotSupportedException to a JsonException could make it a teensy bit harder to figure out you have a misconfiguration in some rare cases. I cannot imagine it breaking a production application. We should absolutely change the exception. type.

**Comment by eiriktsarpalis** [comment]
> I still suspect that fixing System.Text.Json to throw a JsonException instead of a NotSupportedException specifically for this polymorphic deserialization-only case will break absolutely no one and fix a lot of apps.

In simplified terms, what the polymorphic deserializer does is read the type discriminator, look up a type/converter corresponding to it and then hand over control to that converter. There are dozens of reasons why calling that other converter could result in an NSE, mostly related to configuration errors. Superficially, it may appear as if a particular JSON payload prompted the error and must therefore be surfaced as a 4xx, but it truly is a matter of misconfiguration. I don't think that the polymorphic converter should be made to catch and wrap arbitrary NSE's, especially if we agree that it shouldn't happen when the same failure mode occurs in monomorphic contracts.

I maintain that the issue fundamentally is how we handle payloads that don't specify a discriminator. Today this is hardcoded to using the contract for the base type, but we should consider introducing a mode that rejects such payloads altogether, presumably via a flag on `JsonPolymorphicAttribute`.

**Comment by halter73** [comment]
> I don't think that the polymorphic converter should be made to catch and wrap arbitrary NSE's, especially if we agree that it shouldn't happen when the same failure mode occurs in monomorphic contracts.

A flag on JsonPolymorphicAttribute seems like it might be helpful as a component of this fix as long as the broader change fixes this scenario without any code changes to the ASP.NET Core app. I see no reason that an app using the JsonPolymorphicAttribute today should get a NotSupportedException when trying to deserialize a payload without a type discriminator rather than a JsonException.

I accept your contention that where we throw the NotSupportedException today makes it inconvenient to turn this into a JsonException. That makes sense to me. I figure if it were trivial, we would have probably already fixed this bug, because it's clearly undesirable behavior. If we have to catch the NotSupportedException in SystemTextJsonInputFormatter, we will, but I hope we can agree that it's only working around the issue in System.Text.Json.

I don't want the polymorphic converter should be made to catch and wrap arbitrary NSE's, but I do want to either prevent or wrap this one. Why should we have to do it at a higher layer where we have less information? Should we catch all NotSupportedExceptions from System.Text.Json? Do we at least have something other than the exception message to determine whether or not we're running into this particular issue?

**Comment by eiriktsarpalis** [comment]
I suspect we might be talking slightly past each other, so let me see if I can clear things up with another example:

```C#
JsonSerializer.Deserialize<Base>("{}"); // Success
JsonSerializer.Deserialize<Base>("""{"$type":"derived"}"""); // InvalidOperationException

[JsonDerivedType(typeof(Derived), "derived")]
public class Base
{
    public int X { get; set; }
}

public class Derived : Base
{
    [JsonPropertyName("X")]
    public int Y { get; set; }
}
```

In the above, the second payload will prompt an InvalidOperationException because the property name of the derived type is conflicting with the property of the base type. Again, we have a similar situation where a configuration error only gets surfaced once the appropriate discriminator has been specified. Should the exception similarly be intercepted and wrapped in a JsonException? Or is it that we want to special case this specific failure mode?

> Why should we have to do it at a higher layer where we have less information?

Do be very clear, I don't think ASP.NET should have to do anything here. Surfacing a 5xx error is appropriate given that type discriminators are optional. We can have a conversation about enabling a mode where they become required (and therefore prompting a JsonException), so we should just move this issue to runtime to track that.

**Comment by ilya-scale** [comment]
Would it be possible e.g. instead of a flag to automatically detect that the base class cannot be constructed (because it is abstract) and then make the discriminator required?

We can have a flag to always force set it as required as well, but there probably not be a need for that since it is obvious in case of an abstract base class it is what the user intends.

**Comment by halter73** [comment]
> We _could_ special case it for that very specific scenario

I think we should. If there's a clean way to fix this specific scenario without special casing it, that would be even better of course. But if that's the only solution, that's far better than doing nothing.

> , however consider this inverted example:
> ```csharp
> [JsonDerivedType(typeof(Derived), "derived")]
> public record Base;
> 
> public abstract record Derived : Base;
> ```
> In this case the type discriminator is unambiguously optional, however once I specify `"derived"` on the payload the serializer will dispatch to the converter for `Derived`, prompting again a `NotSupportedException` since the particular type does not support deserialization.

I think we should continue throwing a `NotSupportedException` in this case. However, if the fix required that we change this uncommon example to improperly throw a `JsonException` instead, while not ideal, it would be far better than the alternative of throwing a `NotSupportedException`. I think a type discriminator missing when deserializing an abstract base type is the far, far more common scenario.

> I still think that this is best addressed via a flag explicitly marking the type discriminator as required at the polymorphic converter layer.

I think the might could still be somewhat useful for cases where the base type is not abstract for some reason, but you still don't want to deserialize directly to it unless explicitly specified. I still see absolutely no reason why the code in an ASP.NET Core application as described in the issue shouldn't automatically get this behavior considering the base type is abstract. We already have all the information we need to know the type discriminator is required in this case.

I hate asking for redundant configuration that won't fix any existing applications and that most people probably won't know to use when we could just fix the behavior.

> An abstract type might still be deserializable if the user has applied [contract customization](https://learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/custom-contracts). It's not something that the polymorphic converter can determine unless it interrogates the underlying converter for the particular runtime type.

If you think this is a real scenario, we should add a test for this and make sure we don't regress the behavior. However, when there is no contract customization and we're doing bog standard polymorphic deserialization of an abstract base type, we should still throw a `JsonException`.

**Comment by zachsaw-lmg** [comment]
I'd like to add to the discussion by echoing @ilya-scale and @halter73:  
a missing type discriminator for a polymorphic request is fundamentally a 400 Bad Request, not a 500 Internal Server Error.

1. Client vs. Server Errors  
   - A missing discriminator is malformed input.  
   - Classifying it as a 500 triggers pager alerts, causes fatigue, and hides real server problems.

2. Industry Convention  
   - Spring Boot and most frameworks map JSON-syntax or schema failures to 400.  
   - ASP.NET Core’s behavior here is an outlier and violates the principle of least surprise. Imagine waking up at 3am by a prod alert thanks to this little surprise!

3. False Dichotomy  
   - “Deserialization ≠ validation” ignores the fact that model binding is part of input validation.  
   - When deserialization fails due to malformed JSON, it is, practically, a validation error.

4. Why Exception Filters Aren’t Enough  
   - Catching all `JsonException` or `NotSupportedException` forces every team to implement global handlers.  
   - It mixes concerns and risks catching exceptions from application logic.

5. The Right Fix  
   - In the specific case of deserializing an abstract polymorphic type with no discriminator, throw a `JsonException` instead of `NotSupportedException`.  
   - This change alone will let the built-in formatter produce a 400 without extra filters.

**Comment by PaulVrugt** [comment]
How on earth is this a discussion at all? Bad input = 400, there is no discussion to be had here. For consumers, it doesn't matter if and how the api backend has implemented polymorphism. When a user fails to supply a required property, the api should return a 400 stating the consumer messed up their input, not provide a cryptic 500 message, which implies the server is broken

**Comment by iPilot** [comment]
> The type discriminator is not a required field though. Not specifying one makes the serializer default to the base type.

I thought `JsonUnknownDerivedTypeHandling.FailSerialization` from polymorphic configuration is responsible for this and I read it like "throw when payload couldn't be mapped to a class, do not try to instantiate anything". 2 other members cover different behavior.
